cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/autoinstall/HubPackage.java

```diff
@@ -0,0 +1,254 @@
+    Spec spec = GSON.fromJson(HttpClients.doGetAsString(specURL), Spec.class);
+    for (Spec.Action action: spec.getActions()) {
+      // Get plugin jar and json names under one_step_deploy_plugin action in the spec
+      // See https://cdap.atlassian.net/wiki/spaces/DOCS/pages/554401840/Hub+API?src=search#one_step_deploy_plugin
+      if (!action.getType().equals("one_step_deploy_plugin")) {
+        continue;
+      }
+
+      String configFilename = action.getConfigFilename();
+      if (configFilename == null) {
+        LOG.error("Config not found for plugin {}", name);
+        continue;
+      }
+      URL configURL = new URL(url, Joiner.on("/").join(Arrays.asList("/v2/packages", name, version, configFilename)));
+      // Download plugin json from hub
+      JsonObject jsonObj = GSON.fromJson(HttpClients.doGetAsString(configURL), JsonObject.class);
+      List<String> parents = GSON.fromJson(jsonObj.get("parents").toString(), new TypeToken<ArrayList<String>>() {
+      }.getType());
+
+      String jarName = action.getJarName();
+      if (jarName == null) {
+        LOG.error("Jar not found for plugin {}", name);
+        continue;
+      }
+      URL jarURL = new URL(url, Joiner.on("/").join(Arrays.asList("/v2/packages", name, version, jarName)));
+      // Download plugin jar from hub
+      byte[] jar = HttpClients.doGetAsBytes(jarURL);
+
+      final File destination = File.createTempFile("artifact-", ".jar", tmpDir);
+      try {
+        Files.write(jar, destination);
+        io.cdap.cdap.proto.id.ArtifactId artifactId = NamespaceId.DEFAULT.artifact(name, version);
+
+
+        Set<ArtifactRange> parentArtifacts = new HashSet<>();
+        for (String parent : parents) {
+          try {
+            // try parsing it as a namespaced range like system:cdap-data-pipeline[6.3 1.1,7.0.0)
+            parentArtifacts.add(ArtifactRanges.parseArtifactRange(parent));
+          } catch (InvalidArtifactRangeException e) {
+            // if this failed, try parsing as a non-namespaced range like cdap-data-pipeline[6.3 1.1,7.0.0)
+            parentArtifacts.add(ArtifactRanges.parseArtifactRange(NamespaceId.DEFAULT.getNamespace(), parent));
+          }
+        }
+
+        // add the artifact to the repo
+        Id.Artifact artifact = Id.Artifact.fromEntityId(artifactId);
+        try {
+          artifactRepository.addArtifact(artifact,
+                                         destination, parentArtifacts, ImmutableSet.of());
+        } catch (ArtifactAlreadyExistsException e) {
+          LOG.debug("Artifact artifact {}-{} already exists", name, version);
+        }
+
+        Map<String, String> properties = GSON.fromJson(jsonObj.get("properties"), new TypeToken<Map<String, String>>() {
+        }.getType());
+        artifactRepository.writeArtifactProperties(Id.Artifact.fromEntityId(artifactId), properties);
+      } finally {
+        if (destination.exists() && !destination.delete()) {
```

Comment:
Use `Files.deleteIfExists()` instead to simplify the logic.

Reply:
done